extensions/smb/PutSmb.cpp

```diff
@@ -0,0 +1,94 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PutSmb.h"
+#include "utils/gsl.h"
+#include "utils/ProcessorConfigUtils.h"
+#include "utils/OsUtils.h"
+#include "utils/file/FileWriterCallback.h"
+#include "core/Resource.h"
+
+namespace org::apache::nifi::minifi::extensions::smb {
+
+void PutSmb::initialize() {
+  setSupportedProperties(Properties);
+  setSupportedRelationships(Relationships);
+}
+
+void PutSmb::onSchedule(core::ProcessContext* context, core::ProcessSessionFactory*) {
+  gsl_Expects(context);
+  smb_connection_controller_service_ = SmbConnectionControllerService::getFromProperty(*context, PutSmb::ConnectionControllerService);
+  create_missing_dirs_ = context->getProperty<bool>(PutSmb::CreateMissingDirectories).value_or(true);
+  conflict_resolution_strategy_ = utils::parseEnumProperty<FileExistsResolutionStrategy>(*context, ConflictResolution);
```

Comment:
Did this start throwing on invalid values, instead of proceeding with a warning? Would it make sense to warn and fallback to a sensible default in these cases?

Reply:
Yeah, Im a bit torn on it. I think we should definetly throw on new processors like PutSmb, but since I also changed it in PutFile im not sure if we should do this.

Imho we should throw everywhere if we encounter invalid configs like this, or if thats too drastic we should change them to throw after 1.0 release.

In the meantime I've added a WIP commit that reverts PutFile to the older behaviour. https://github.com/apache/nifi-minifi-cpp/pull/1634/commits/0dd3adf71b517621274bc06891d29f259f63f4d1

If you like it I can add some tests or maybe we could implement monoadic operators to nonstd::expected aswell which would make them easier to use.
Or if we take a bit more aggressive stance on this I can just revert this commit and let it throw...

Reply:
I guess you're right: it's not ideal that we're breaking some invalid configs, but the users can probably handle some necessary fixing of their configs after an upgrade. Let's keep the original throwing version, and not forget to include this info in the next changelog / release notes.

Reply:
👍 reverted to the throwing variant in https://github.com/apache/nifi-minifi-cpp/pull/1634/commits/325093ae7f106771368191a61895a3bc69cabf22

Reply:
There are some monadic operations implemented on `nonstd::expected` in our `libminifi/include/utils/expected.h`. Are we missing something important for this use case?